File: Src/Workspaces/VisualBasic/Formatting/Engine/Trivia/TriviaFormatter.vb
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports Microsoft.VisualBasic
Imports Roslyn.Compilers.Internal
Imports Roslyn.Compilers.VisualBasic
Imports Roslyn.Services.VisualBasic.Extensions
Imports Roslyn.Services.Formatting

Namespace Roslyn.Services.VisualBasic.Formatting
    Partial Friend Class TriviaFormatter

#Region "Caches"
        Private Shared ReadOnly LineContinuationCache As SyntaxTrivia
        Private Shared ReadOnly SpaceCache() As String
        Private Shared ReadOnly Pool As TriviaListPool
#End Region

        Private ReadOnly context As FormattingContext
        Private ReadOnly token1 As SyntaxToken
        Private ReadOnly token2 As SyntaxToken

        Private ReadOnly lineBreaks As Integer
        Private ReadOnly spaces As Integer

        Shared Sub New()
            SpaceCache = New String(19) {}
            For i As Integer = 0 To 19
                SpaceCache(i) = New String(" "c, i)
            Next i

            LineContinuationCache = Syntax.LineContinuationTrivia("_" + vbCrLf)

            Pool = New TriviaListPool()
        End Sub

        Public Sub New(context As FormattingContext, token1 As SyntaxToken, token2 As SyntaxToken, lineBreaks As Integer, spaces As Integer)
            Contract.ThrowIfNull(context)
            Contract.ThrowIfFalse(lineBreaks >= 0)
            Contract.ThrowIfFalse(spaces >= 0)

            Me.context = context

            Me.token1 = token1
            Me.token2 = token2

            Me.lineBreaks = lineBreaks
            Me.spaces = spaces
        End Sub

        Public Function FormatToSyntaxTriviaList() As List(Of SyntaxTrivia)
            Dim list = TriviaListPool.Allocate()

            FormatAndAppendToSyntaxTriviaList(list)
            Return TriviaListPool.ReturnAndFree(list)
        End Function

        Public Function FormatToString() As String
            ' optimized for common case
            ' used pool to get short live heap objects
            Dim builder = StringBuilderPool.Allocate()
            Dim list = TriviaListPool.Allocate()

            FormatAndAppendToSyntaxTriviaList(list)
            For i As Integer = 0 To list.Count - 1
                builder.Append(list(i).GetFullText())
            Next i

            TriviaListPool.Free(list)
            Return StringBuilderPool.ReturnAndFree(builder)
        End Function

        Private Sub FormatAndAppendToSyntaxTriviaList(triviaList As List(Of SyntaxTrivia))
            Dim currentLineBreaks As Integer

            Dim buffer = TriviaListPool.Allocate()
            Dim triviaBuilder = ProcessTriviaOnEachLine(buffer, triviaList, currentLineBreaks)

            ' single line trivia case
            If Me.lineBreaks = 0 AndAlso currentLineBreaks = 0 Then
                Dim initialColumn = Me.context.TreeInfo.GetColumnOfToken(token1, Me.context.Options.TabSize) + token1.Width()
                triviaBuilder.CommitBetweenTokens(initialColumn, triviaList)
                Return
            End If

            ' if there was trivia left over, don't bother to add indentation. probably code is
            ' malformed anyway.
            Dim beginningOfNewLine = currentLineBreaks <> 0 OrElse IsFirstTriviaInTreeOrAfterStatementTerminator()
            If triviaBuilder.CommitLeftOver(beginningOfNewLine, triviaList, currentLineBreaks) Then
                Return
            End If

            AppendLineBreakTrivia(currentLineBreaks, triviaList)

            ' remove indentation if we are formatting "<token>\r\n .. <some noisy trivia such as line continuation>\r\n .. <statement termination token>"
            If Me.token2.Kind = SyntaxKind.StatementTerminatorToken Then
                Return
            End If

            Dim indentationString = Me.spaces.CreateIndentationString(Me.context.Options.UseTab, Me.context.Options.TabSize)
            AppendWhitespaceTrivia(indentationString, triviaList)

            TriviaListPool.Free(buffer)
        End Sub

        Private Sub AppendLineBreakTrivia(currentLineBreaks As Integer, list As List(Of SyntaxTrivia))
            If currentLineBreaks >= Me.lineBreaks Then
                Return
            End If

            ' by default, prepend extra new lines asked rather than append.
            Dim tempList = TriviaListPool.Allocate()
            For i As Integer = currentLineBreaks To Me.lineBreaks - 2
                tempList.Add(Syntax.CarriageReturnLineFeed)
            Next i

            list.InsertRange(0, tempList)
            TriviaListPool.Free(tempList)
        End Sub

        Private Function ProcessTriviaOnEachLine(buffer As List(Of SyntaxTrivia), triviaList As List(Of SyntaxTrivia), ByRef currentLineBreaks As Integer) As TriviaLineBuilder
            Dim triviaBuilder = New TriviaLineBuilder(Me.context, buffer)

            ' initialize
            currentLineBreaks = 0

            For Each currentTrivia In GetTriviaBetweenTokens()

                ' trivia list we have is between two tokens. so trailing trivia belongs to previous token will not
                ' start from new line.
                Dim beginningOfNewLine = currentLineBreaks <> 0 OrElse IsFirstTriviaInTreeOrAfterStatementTerminator()

                If currentTrivia.IsElastic OrElse
                   currentTrivia.Kind = SyntaxKind.EndOfLineTrivia OrElse
                   currentTrivia.Kind = SyntaxKind.ImplicitLineContinuationTrivia Then
                    currentLineBreaks += triviaBuilder.CommitLines(beginningOfNewLine, triviaList)

                    triviaList.Add(Syntax.CarriageReturnLineFeed)
                    currentLineBreaks += 1

                    triviaBuilder.Reset()
                    Continue For
                ElseIf currentTrivia.Kind = SyntaxKind.LineContinuationTrivia OrElse
                       currentTrivia.Kind = SyntaxKind.DocumentationComment OrElse
                       currentTrivia.Kind.IsPreprocessorDirective() OrElse
                       currentTrivia.Kind = SyntaxKind.DisabledTextTrivia Then
                    triviaBuilder.Add(currentTrivia)
                    currentLineBreaks += triviaBuilder.CommitLines(beginningOfNewLine, triviaList)

                    triviaBuilder.Reset()
                    Continue For
                End If

                triviaBuilder.Add(currentTrivia)
            Next currentTrivia

            Return triviaBuilder
        End Function

        Private Shared Sub AppendWhitespaceTrivia(whitespaceString As String, list As List(Of SyntaxTrivia))
            If String.IsNullOrEmpty(whitespaceString) Then
                Return
            End If

            list.Add(Syntax.WhitespaceTrivia(whitespaceString))
        End Sub

        Private Function IsFirstTriviaInTreeOrAfterStatementTerminator() As Boolean
            Return Me.token1.Kind = SyntaxKind.None OrElse Me.token1.Kind = SyntaxKind.StatementTerminatorToken
        End Function

        Private Shared Function GetSpaces(space As Integer) As String
            If space >= 0 AndAlso space < 20 Then
                Return SpaceCache(space)
            End If

            Return New String(" "c, space)
        End Function

        Private Function GetTriviaBetweenTokens() As IEnumerable(Of SyntaxTrivia)
            Dim list = New List(Of SyntaxTrivia)

            If Me.token1.Kind <> SyntaxKind.None AndAlso Me.token1.HasTrailingTrivia Then
                For i As Integer = 0 To Me.token1.TrailingTrivia.Count - 1
                    list.Add(Me.token1.TrailingTrivia(i))
                Next i
            End If

            If Me.token2.Kind <> SyntaxKind.None AndAlso Me.token2.HasLeadingTrivia Then
                For i As Integer = 0 To Me.token2.LeadingTrivia.Count - 1
                    list.Add(Me.token2.LeadingTrivia(i))
                Next i
            End If

            Return list
        End Function

        Public Shared Function ContainsSkippedTokensOrText(list As TriviaList) As Boolean
            For i As Integer = 0 To list.Count - 1
                Dim trivia = list(i)

                If trivia.Kind = SyntaxKind.SkippedTokens Then
                    Return True
                End If
            Next i

            Return False
        End Function

        Friend Shared Function ShouldFormatTriviaOnSingleLine(triviaList As TriviaList) As Boolean

            For i As Integer = 0 To triviaList.Count - 1
                Dim trivia = triviaList(i)

                Contract.ThrowIfTrue(trivia.Kind = SyntaxKind.EndOfLineTrivia)
                Contract.ThrowIfTrue(trivia.Kind = SyntaxKind.SkippedTokens)

                ' if it contains elastic trivia. always format
                If trivia.IsElastic Then
                    Return True
                End If

                If trivia.Kind = SyntaxKind.WhitespaceTrivia Then
                    Debug.Assert(trivia.GetText() = trivia.GetFullText())
                    Dim text = trivia.GetText()
                    If text.IndexOf(vbTab) >= 0 Then
                        Return True
                    End If
                End If

                If trivia.Kind = SyntaxKind.CommentTrivia Then
                    ' need to format if comment sit right after a token
                    Return i = 0
                End If

                If trivia.Kind = SyntaxKind.DocumentationComment Then
                    Return False
                End If

                If trivia.Kind = SyntaxKind.RegionDirective OrElse trivia.Kind = SyntaxKind.EndRegionDirective OrElse trivia.Kind.IsPreprocessorDirective() Then
                    Return False
                End If
            Next i

            Return True
        End Function

        Public Shared Function ShouldFormatTriviaOnMultipleLines(options As FormattingOptions,
                                                                 beginningOfNewLine As Boolean,
                                                                 desiredIndentation As Integer,
                                                                 list As TriviaList) As Boolean
            Return MultiLineAnalyzer.ShouldFormat(options, beginningOfNewLine, desiredIndentation, list)
        End Function
    End Class
End Namespace

File: Src/Workspaces/CSharp/Formatting/Engine/Trivia/TriviaFormatter.cs
using System.Collections.Generic;
using System.Diagnostics;
using Roslyn.Compilers.CSharp;
using Roslyn.Services.CSharp.Extensions;
using Roslyn.Services.Formatting;
using Roslyn.Utilities;

namespace Roslyn.Services.CSharp.Formatting
{
    internal partial class TriviaFormatter
    {
        #region Caches
        private static readonly string[] SpaceCache;
        #endregion

        private readonly FormattingContext context;
        private readonly SyntaxToken token1;
        private readonly SyntaxToken token2;

        private readonly int lineBreaks;
        private readonly int spaces;

        static TriviaFormatter()
        {
            SpaceCache = new string[20];
            for (int i = 0; i < 20; i++)
            {
                SpaceCache[i] = new string(' ', i);
            }
        }

        public TriviaFormatter(FormattingContext context, SyntaxToken token1, SyntaxToken token2, int lineBreaks, int spaces)
        {
            Contract.ThrowIfNull(context);
            Contract.ThrowIfFalse(lineBreaks >= 0);
            Contract.ThrowIfFalse(spaces >= 0);

            this.context = context;

            this.token1 = token1;
            this.token2 = token2;

            this.lineBreaks = lineBreaks;
            this.spaces = spaces;
        }

        public List<SyntaxTrivia> FormatToSyntaxTriviaList()
        {
            var list = TriviaListPool.Allocate();

            FormatAndAppendToSyntaxTriviaList(list);
            return TriviaListPool.ReturnAndFree(list);
        }

        public string FormatToString()
        {
            // optimized for common case
            // used pool to get short live heap objects
            var builder = StringBuilderPool.Allocate();
            var list = TriviaListPool.Allocate();

            FormatAndAppendToSyntaxTriviaList(list);

            for (int i = 0; i < list.Count; i++)
            {
                builder.Append(list[i].GetFullText());
            }

            TriviaListPool.Free(list);
            return StringBuilderPool.ReturnAndFree(builder);
        }

        private void FormatAndAppendToSyntaxTriviaList(List<SyntaxTrivia> triviaList)
        {
            int currentLineBreaks;

            var buffer = TriviaListPool.Allocate();
            var triviaBuilder = ProcessTriviaOnEachLine(buffer, triviaList, out currentLineBreaks);

            // single line trivia case
            if (this.lineBreaks == 0 && currentLineBreaks == 0)
            {
                var initialColumn = this.context.TreeInfo.GetColumnOfToken(token1, this.context.Options.TabSize) + token1.Width();
                triviaBuilder.CommitBetweenTokens(initialColumn, triviaList);
                return;
            }

            // multi line trivia case
            var beginningOfNewLine = currentLineBreaks != 0;
            currentLineBreaks += triviaBuilder.CommitLeftOver(beginningOfNewLine, this.spaces, triviaList);

            AppendLineBreakTrivia(currentLineBreaks, triviaList);

            var indentationString = this.spaces.CreateIndentationString(this.context.Options.UseTab, this.context.Options.TabSize);

            AppendWhitespaceTrivia(indentationString, triviaList);
            TriviaListPool.Free(buffer);
        }

        private void AppendLineBreakTrivia(int currentLineBreaks, List<SyntaxTrivia> triviaList)
        {
            if (currentLineBreaks >= this.lineBreaks)
            {
                return;
            }

            // by default, prepend extra new lines asked rather than append.
            var tempList = TriviaListPool.Allocate();
            for (int i = currentLineBreaks; i < this.lineBreaks; i++)
            {
                tempList.Add(Syntax.CarriageReturnLineFeed);
            }

            triviaList.InsertRange(0, tempList);
            TriviaListPool.Free(tempList);
        }

        private TriviaLineBuilder ProcessTriviaOnEachLine(List<SyntaxTrivia> buffer, List<SyntaxTrivia> triviaList, out int currentLineBreaks)
        {
            var triviaBuilder = new TriviaLineBuilder(this.context, buffer);

            // initialize
            currentLineBreaks = 0;

            var treatElasticAsNewLine = true;
            foreach (var trivia in GetTriviaBetweenTokens())
            {
                // trivia list we have is between two tokens. so trailing trivia belongs to previous token will not
                // start from new line.
                var beginningOfNewLine = currentLineBreaks != 0 || IsFirstTriviaInTree();

                if (trivia.IsElastic && treatElasticAsNewLine)
                {
                    treatElasticAsNewLine = false;
                    CommitLinesAndAddLineBreak(beginningOfNewLine, triviaList, ref currentLineBreaks, ref triviaBuilder);
                    continue;
                }

                if (trivia.Kind == SyntaxKind.EndOfLineTrivia)
                {
                    treatElasticAsNewLine = false;
                    CommitLinesAndAddLineBreak(beginningOfNewLine, triviaList, ref currentLineBreaks, ref triviaBuilder);
                    continue;
                }

                if (trivia.Kind == SyntaxKind.SingleLineCommentTrivia ||
                    trivia.Kind == SyntaxKind.MultiLineCommentTrivia)
                {
                    treatElasticAsNewLine = true;
                    triviaBuilder.Add(trivia);
                    continue;
                }

                if (trivia.Kind == SyntaxKind.DocumentationComment ||
                    trivia.Kind == SyntaxKind.DisabledTextTrivia ||
                    trivia.Kind.IsPreprocessorDirective())
                {
                    treatElasticAsNewLine = true;

                    if (trivia.Kind == SyntaxKind.DocumentationComment && !IsSingleLineDocumentComment(trivia))
                    {
                        triviaBuilder.Add(trivia);
                        continue;
                    }

                    CommitLines(beginningOfNewLine, triviaList, trivia, ref currentLineBreaks, ref triviaBuilder);
                    continue;
                }

                treatElasticAsNewLine = false;
                triviaBuilder.Add(trivia);
            }

            return triviaBuilder;
        }

        private static void CommitLines(
            bool beginningOfNewLine,
            List<SyntaxTrivia> triviaList,
            SyntaxTrivia trivia,
            ref int currentLineBreaks,
            ref TriviaLineBuilder triviaBuilder)
        {
            triviaBuilder.Add(trivia);
            currentLineBreaks += triviaBuilder.CommitLines(beginningOfNewLine, triviaList);

            triviaBuilder.Reset();
        }

        private static void CommitLinesAndAddLineBreak(
            bool beginningOfNewLine,
            List<SyntaxTrivia> triviaList,
            ref int currentLineBreaks,
            ref TriviaLineBuilder triviaBuilder)
        {
            currentLineBreaks += triviaBuilder.CommitLines(beginningOfNewLine, triviaList);

            triviaList.Add(Syntax.CarriageReturnLineFeed);
            currentLineBreaks++;

            triviaBuilder.Reset();
        }

        private static bool IsSingleLineDocumentComment(SyntaxTrivia trivia)
        {
            if (trivia.Kind != SyntaxKind.DocumentationComment)
            {
                return false;
            }

            // try to find first doc comment exterior trivia
            var xmlComment = (DocumentationCommentSyntax)trivia.GetStructure();
            foreach (var token in xmlComment.DescendantTokens())
            {
                foreach (var xmlTrivia in token.LeadingTrivia)
                {
                    if (xmlTrivia.Kind == SyntaxKind.DocumentationCommentExteriorTrivia)
                    {
                        return xmlTrivia.GetText().LastIndexOf("///") >= 0;
                    }
                }
            }

            return false;
        }

        private static void AppendWhitespaceTrivia(string indentationString, List<SyntaxTrivia> triviaList)
        {
            if (string.IsNullOrEmpty(indentationString))
            {
                return;
            }

            triviaList.Add(Syntax.Whitespace(indentationString, elastic: false));
        }

        private bool IsFirstTriviaInTree()
        {
            return this.token1.Kind == SyntaxKind.None;
        }

        private static string GetSpaces(int space)
        {
            if (space >= 0 && space < 20)
            {
                return SpaceCache[space];
            }

            return new string(' ', space);
        }

        private IEnumerable<SyntaxTrivia> GetTriviaBetweenTokens()
        {
            if (this.token1.Kind != SyntaxKind.None && this.token1.HasTrailingTrivia)
            {
                for (int i = 0; i < this.token1.TrailingTrivia.Count; i++)
                {
                    yield return this.token1.TrailingTrivia[i];
                }
            }

            if (this.token2.Kind != SyntaxKind.None && this.token2.HasLeadingTrivia)
            {
                for (int i = 0; i < this.token2.LeadingTrivia.Count; i++)
                {
                    yield return this.token2.LeadingTrivia[i];
                }
            }
        }

        public static bool ContainsSkippedTokensOrText(TriviaList list)
        {
            for (int i = 0; i < list.Count; i++)
            {
                var trivia = list[i];

                if (trivia.Kind == SyntaxKind.SkippedTokens ||
                    trivia.Kind == SyntaxKind.PreprocessingMessageTrivia)
                {
                    return true;
                }
            }

            return false;
        }

        public static bool ShouldFormatTriviaOnSingleLine(TriviaList list)
        {
            for (int i = 0; i < list.Count; i++)
            {
                var trivia = list[i];

                Contract.ThrowIfTrue(trivia.Kind == SyntaxKind.EndOfLineTrivia);
                Contract.ThrowIfTrue(trivia.Kind == SyntaxKind.SkippedTokens);
                Contract.ThrowIfTrue(trivia.Kind == SyntaxKind.PreprocessingMessageTrivia);

                // if it contains elastic trivia, always format
                if (trivia.IsElastic)
                {
                    return true;
                }

                if (trivia.Kind == SyntaxKind.WhitespaceTrivia)
                {
                    Debug.Assert(trivia.GetText() == trivia.GetFullText());
                    var text = trivia.GetText();
                    if (text.IndexOf('\t') >= 0)
                    {
                        return true;
                    }
                }

                if (trivia.Kind == SyntaxKind.SingleLineCommentTrivia)
                {
                    return false;
                }

                if (trivia.Kind == SyntaxKind.MultiLineCommentTrivia)
                {
                    // we don't touch space between two tokens on a single line that contains
                    // multiline comments between them
                    return false;
                }

                if (trivia.Kind == SyntaxKind.DocumentationComment)
                {
                    if (IsSingleLineDocumentComment(trivia))
                    {
                        return false;
                    }

                    // we don't touch space between two tokens on a single line that contains
                    // multiline doc comments between them
                    return false;
                }

                if (trivia.Kind == SyntaxKind.RegionDirective ||
                    trivia.Kind == SyntaxKind.EndRegionDirective ||
                    trivia.Kind.IsPreprocessorDirective())
                {
                    return false;
                }
            }

            return true;
        }

        public static bool ShouldFormatTriviaOnMultipleLines(FormattingOptions options, bool firstTriviaInTree, int desiredIndentation, TriviaList list)
        {
            return MultiLineAnalyzer.ShouldFormat(options, firstTriviaInTree, desiredIndentation, list);
        }
    }
}
